Fix --cache set only if with --write or no-different

Milly opened the pull request:
## Description

Do not save formatted cache if without `--write` option and the file is modifies by prettier.

Before fixed, always save formatted cache if with `--cache` option.
Therefore, unformatted files will never be formatted.

I think this is a bug.

### Before fixed:
```sh
> prettier --cache foo.js
# Show formatted contents of `foo.js`.
# Then the cache is created and `foo.js` is flagged as already formatted.

> prettier --cache --write foo.js
foo.js 2ms (cached)
# "... (cached)" means that the file is already formatted and do nothing this time.
```

### After fixed with this PR:
```sh
> prettier --cache foo.js
# Show formatted contents of `foo.js`.

> prettier --cache --write foo.js
foo.js 2ms
# `foo.js` has been formatted and overwritten.
```

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

Milly commented:
> Why don't we update cache when file actually successfully written?

Doing so adds too complexity due to `performanceTestFlag` condition.
I made a fix to reduce the diff.

### Conditions to be cached

| Options | File needs to be formatted | Should be cached |
|--|--|--|
| `--cache` | Yes | No |
| `--cache` | No | Yes |
| `--cache --write` | Yes | Yes |
| `--cache --write` | No | Yes |

Milly commented:
Should we not write the cache file if `performanceTestFlag ` is specified?
I don't understand.

fisker commented:
We don't need care about  `performanceTestFlag` that's for debugging.

Milly commented:
> Why don't we update cache when file actually successfully written?

I fixed it.

Milly commented:
This fixes #13015

Milly commented:
I found an additional problem.
Fixing now.

```sh
> prettier --cache --write foo.js
foo.js 2ms
# Cache created.

> vi foo.js
# Edit to unformatted.

> prettier --list-different --cache foo.js
foo.js
# Cache is updated because the file entry is already exist in cache and update all.

> prettier --list-different --cache foo.js
# Expected `foo.js` will be output, but nothing.
```

Milly commented:
Fixed it.

Milly commented:
refs #12800

sosukesuzuki commented on src/cli/format.js:
```diff
@@ -405,10 +405,11 @@ async function formatFiles(context) {
     const isDifferent = output !== input;
 
+    // Only set formatted cache if with --write or no-different.
+    let isCacheSet = context.argv.write || !isDifferent;
```
?
```suggestion
    let shouldSetCache = context.argv.write || !isDifferent;
```

sosukesuzuki reviewed (commented):
Looks good, please check one comment. And please add changelog.

darekkay reviewed (commented):
I think this PR is just a workaround. ESLint [handles](https://github.com/prettier/prettier/issues/13015#issuecomment-1168776169) this case correctly without deactivating the cache.

fisker commented on src/cli/format.js:
```diff
@@ -460,6 +464,12 @@ async function formatFiles(context) {
+    if (shouldSetCache) {
+      formatResultsCache?.setFormatResultsCache(filename, options);
+    } else {
+      formatResultsCache?.removeFormatResultsCache(filename);
```
Can we just do nothing if the file not written? Will it cause problem?

Milly replied:
@tksst says, cause problem.
https://github.com/prettier/prettier/pull/13016#issuecomment-1214311862

I'll revert changes.

fisker commented on tests/integration/run-prettier.js:
```diff
@@ -40,12 +40,21 @@ async function run(dir, args, options) {
     .spyOn(fs.promises, "writeFile")
-    // eslint-disable-next-line require-await
-    .mockImplementation(async (filename, content) => {
+    .mockImplementation(async function(filename, content) {
```
This is good as it is, the problem is in `next` branch, we don't relay on `mockImplementation`, do you think it's possible to use a simple option like `{mockWriteFileError: 'error message'}`, so we can pass it to the worker? Otherwise this will be hard when we merge into the `next` branch

https://github.com/prettier/prettier/blob/166138cfddcc3f5ecab9b366e35270d0c3f79d0f/tests/integration/cli-worker.js#L12-L17

fisker replied:
Or maybe `{mockWriteFileError: {'a.js': 'error message'}}` in this case.

fisker commented on tests/integration/run-prettier.js:
```diff
@@ -40,20 +40,15 @@ async function run(dir, args, options) {
-        }
+    // eslint-disable-next-line require-await
+    .mockImplementation(async (filename, content) => {
+      const error = options.mockWriteFileErrors?.[filename];
```
Is the filename an absolute path? `path.relative()`?

Milly replied:
as is?

Milly replied:
I think this is fine.

fisker commented on src/cli/format-results-cache.js:
```diff
@@ -86,10 +87,22 @@ class FormatResultsCache {
+  /**
+   * @param {string} filePath
+   */
+  removeFormatResultsCache(filePath) {
```
This seems not used.

fisker reviewed (approved):
Thanks for working on this.

tksst commented:
The current code still does not work properly in the case that multiple files are edited at the same time and one or more files are updated in a formatted state.

```shell
$ rm -rf node_modules/.cache
# clear cache

$ node bin/prettier.js --cache --check a.js b.js
Checking formatting...
All matched files use Prettier code style!
# Confirm that these are currently formatted

$ echo "   " >> a.js
# modify to unformatted.

$ echo "const x = 1;" >> b.js
# modify but still formatted.

$ node bin/prettier.js --cache --check a.js b.js
Checking formatting...
[warn] a.js
[warn] Code style issues found in the above file. Forgot to run Prettier?

$ node bin/prettier.js --cache --write a.js b.js
a.js 0ms (cached)
b.js 1ms (cached)
# a.js was cached as formatted!
```

Not only [`setFormatResultsCache`](https://github.com/prettier/prettier/blob/0f19075f5f0bc3ecca0c583a0be699eca157ff06/src/cli/format-results-cache.js#L83) but also [`existsAvailableFormatResultsCache`](https://github.com/prettier/prettier/blob/0f19075f5f0bc3ecca0c583a0be699eca157ff06/src/cli/format-results-cache.js#L61) cache a current state of file.
I would like to suggest [a simple patch to remove cache entries for unformatted files.](https://github.com/tksst/prettier/commit/a65ed45dd302108cf49824c6bd9771a5e885919b)

Milly commented:
- Revert [just do nothing if the file not written](#pullrequestreview-1056678210).
- Added tests for [multipe files updating same time](#issuecomment-1214311862). (Thanks @tksst)
- Cleaned up commit history.
- Rebase to latest main.

fisker commented:
close/reopen to trigger ci.

fisker commented:
It's back..

fisker commented on tests/integration/__tests__/cache.js:
```diff
@@ -227,6 +227,95 @@ describe("--cache option", () => {
       );
     });
 
+    it("re-formats after execution without write.", async () => {
```
I don't think it's a good idea to relay on another `it` call, should format twice in one block. otherwise it.only will fail.

fisker replied:
Ah, seems already did. Sorry.

fisker commented on tests/integration/__tests__/cache.js:
```diff
@@ -227,6 +227,95 @@ describe("--cache option", () => {
     });
 
+    it("re-formats after execution without write.", async () => {
+      await runPrettier(dir, ["--cache", "--cache-strategy", "metadata", "."]);
```
Let's put this call in a function, so we don't need check two places when updating. 

Maybe even better to accept `times` as a argument.

fisker replied:
Or simply reuse an array to store cli arguments.

sosukesuzuki replied:
@fisker I've added `getCliArguments` util function. What do you think about [d653fe1](https://github.com/prettier/prettier/pull/13016/commits/d653fe10a92390102ae97cb43b2efcbb5c3c9140) ?

fisker replied:
It's unreadable... I only ask to make it easier to maintain the second format arguments by putting them together.

sosukesuzuki replied:
Like this?: [9406d54](https://github.com/prettier/prettier/pull/13016/commits/9406d54946d7e80c30b26b66965e7419258955ca)

fisker replied:
Yes.

bryanjtc commented:
Update on this?

fisker commented:
Feel this will be hard to fix conflicts when merge to next branch.

sosukesuzuki commented:
@Milly Thank you for doing this! Good job!

fisker commented:
@sosukesuzuki Good job merging this into `next`!

pspeter3 commented:
Will `--check` use the cache?
